src/sgame/lua/Interpreter.cpp

```diff
@@ -0,0 +1,318 @@
+		( *(int*)( ud ) ) = -1;
+		lua_pushcclosure( L, rocket_pairsaux, 2 );  // uv 1 is __ipairs, uv 2 is ud
+	}
+	else
+	{
+		lua_pushvalue( L, lua_upvalueindex( 1 ) );  // generator
+	}
+	lua_pushvalue( L, 1 );  // state
+	lua_pushnil( L );       // initial value
+	return 3;
+}
+
+// Based off of the LuaPrint in libRocket, which was based off of
+// luaB_print function from Lua's lbaselib.c
+static int LuaPrint( lua_State* L )
+{
+	int n = lua_gettop( L ); /* number of arguments */
+	int i;
+	lua_getglobal( L, "tostring" );
+	std::vector<std::string> string_list;
+	std::string output = "";
+	for ( i = 1; i <= n; ++i )
+	{
+		const char* s;
+		lua_pushvalue( L, -1 ); /* function to be called */
+		lua_pushvalue( L, i );  /* value to print */
+		lua_call( L, 1, 1 );
+		s = lua_tostring( L, -1 ); /* get result */
+		if ( s == NULL ) return luaL_error( L, "'tostring' must return a string to 'print'" );
+		if ( i > 1 ) output += "\t";
+		output += s;
+		lua_pop( L, 1 ); /* pop result */
+	}
+	Log::Notice( output );
+	return 0;
+}
+
+static int LoadUnvModule( lua_State* L )
+{
+	const char* file = luaL_checkstring( L, 1 );
+	if ( !file )
+	{
+		Log::Warn( "invalid call to LoadUnvModule: must pass in a string." );
+		return 0;
+	}
+	if ( !LoadScript( file ) )
+	{
+		Log::Warn( "error loading script %s", file );
+		return 0;
+	}
+	return 1;
+}
+
+// Remove searchers that search the filesystems so we can install our own searcher.
+// Keep the searcher at index 1 because it just looks for cached modules that have
+// already been loaded.
+const char INITIALIZE_SEARCHERS[] = R"(
+table.remove(package.searchers, 2)
+table.remove(package.searchers, 3)
+)";
+
+static void OverrideGlobalLuaFunctions()
+{
+	lua_getglobal( L, "_G" );
+
+	lua_getglobal( L, "next" );
+	lua_pushcclosure( L, rocket_pairs, 1 );
+	lua_setfield( L, -2, "pairs" );
+
+	lua_getglobal( L, "next" );
+	lua_pushcclosure( L, rocket_ipairs, 1 );
+	lua_setfield( L, -2, "ipairs" );
+
+	lua_pushcfunction( L, LuaPrint );
+	lua_setfield( L, -2, "print" );
+
+	LoadCode( INITIALIZE_SEARCHERS, __func__ ) && RunCode();
+	// Add the package loader to the package.loaders table.
+	lua_getglobal( L, "package" );
+	lua_getfield( L, -1, "searchers" );
+
+	if ( lua_isnil( L, -1 ) )
+	{
+		Sys::Drop( "Can't register searcher: package.loaders table does not exist." );
+	}
+
+	lua_pushcfunction( L, LoadUnvModule );
+	lua_rawseti( L, -2, 2 );
+	lua_pop( L, 2 );
+}
+
+void Initialize()
+{
+	if ( L ) return;
+
+	L = luaL_newstate();
+	luaL_openlibs( L );
+	OverrideGlobalLuaFunctions();
+	BG_InitializeLuaConstants( L );
+}
+
+void Shutdown()
+{
+	lua_close( L );
+	L = nullptr;
+}
+
+lua_State* State()
+{
+	return L;
+}
+
+class LuaCommand : Cmd::StaticCmd
+{
+   public:
+	LuaCommand() : Cmd::StaticCmd( "lua", Cmd::SGAME_VM, "exec server side lua" ) {}
+
+	void Run( const Cmd::Args& args ) const
+	{
+		auto usage = [ & ]()
+		{ PrintUsage( args, "\"[-f filename ] | <lua code>\"", "exec server side lua" ); };
```

Comment:
There shouldn't be quotes inside the string

Reply:
Done